content/docs/concepts/options/transforms.md

```diff
@@ -0,0 +1,327 @@
+
+### No resource type
+
+There is no `Resource` object passed to transform functions. Most of the information you could have retrieved from that object is presented on the transform arguments directly.
+
+For example given the following C# transform:
+
+```csharp
+args =>
+{
+    if (args.Resource.GetResourceType() == "random:index/randomString:RandomString")
+    {
+        var resultOpts = CustomResourceOptions.Merge((CustomResourceOptions)args.Options,
+            new CustomResourceOptions {AdditionalSecretOutputs = {"length"}});
+        return new ResourceTransformationResult(resultArgs, resultOpts);
+    }
+    return null;
+}
+```
+
+We can rewrite this to:
+
+```csharp
+async (args, _) =>
+{
+    if (args.Type == "random:index/randomString:RandomString")
+    {
+        var resultOpts = CustomResourceOptions.Merge((CustomResourceOptions)args.Options,
+            new CustomResourceOptions {AdditionalSecretOutputs = {"length"}});
+        return new ResourceTransformResult(resultArgs, resultOpts);
+    }
+    return null;
+}
+```
+
+### No typed resource args class
+
+In the old transform system the transform function was called with the same object that was passed to the resource constructor. This meant that
+in typed languages like Go and C# you could typecast to the typed arguments struct.
+
+```go
+func(_ context.Context, rta *pulumi.ResourceTransformationArgs) *pulumi.ResourceTransformationResult {
+    if rta.Type == "testprovider:index:Random" {
+        var props *RandomArgs
+        if rta.Props == nil {
+            props = &RandomArgs{}
+        } else {
+            props = rta.Props.(*RandomArgs)
+        }
+        props.Length = props.Length.ToOutput().ApplyT(func(v int) int { return v * 2 })
+
+        return &pulumi.ResourceTransformationResult{
+            Props: props,
+            Opts:  rta.Opts,
+        }
+    }
+    return nil
+}
+```
+
+The new transform system works over the wire protocol, allowing it to run for resources created in other
+processes, but it means the properties object you get is closer to the raw protocol than the typed arguments
+you might expect.
+
+In this Go example the old system would have given us a typed `RandomArgs` structure with the length field
+being a `pulumi.IntInput`. The new system is just a map where we have to know the key is `"length"` and the
+value is a `Float64Output`.
+
+You might ask why `Float64Output` instead of at least `IntOutput`. This is because the wire protocol doesn't
+actually have support for integers (being JSON based), so on receiving the untyped properties the transform
+callback can only go on their JSON values and all numbers are 64-bit floats.
+
+```go
+func(_ context.Context, rta *pulumi.XResourceTransformArgs) *pulumi.XResourceTransformResult {
+    if rta.Type == "testprovider:index:Random" {
+        length := rta.Props["length"].(pulumi.Float64Output)
+        rta.Props["length"] = length.ApplyT(func(v float64) float64 { return v * 2 })
```

Comment:
I think it would be great to expand on how to transform the props, like explain that the values are encapsulated as Outputs, that you should expect variations related to secretness etc, and that you transform the outputs as explained elsewhere.

Reply:
A similar example in Python:
```py
def apply_patchforce_annotation(args: ResourceTransformArgs):
    if not args.type_ == "kubernetes:helm.sh/v4:Chart":
        if not 'metadata' in args.props:
            args.props['metadata'] = {}
        if not 'annotations' in args.props['metadata']:
            args.props['metadata']['annotations'] = {}
        args.props['metadata']['annotations']['pulumi.com/patchForce'] = 'true'

    return ResourceTransformResult(
        props=args.props,
        opts=args.opts)

ingresscontroller = kubernetes.helm.v4.Chart(
    "ingresscontroller",
    chart="nginx-ingress",
    namespace=ingress_ns.metadata.name,
    repository_opts=kubernetes.helm.v3.RepositoryOptsArgs(
        repo="https://helm.nginx.com/stable",
    ),
    version="0.14.1",
    opts=pulumi.ResourceOptions(transforms=[apply_patchforce_annotation])
)
```

Reply:
> I think it would be great to expand on how to transform the props, like explain that the values are encapsulated as Outputs, that you should expect variations related to secretness (should you?), and that you transform the outputs as explained elsewhere.

Agree! It's somewhat shown in some of the examples, but I'd like to expand on this as a fast follow.

> Would be awesome to show an example like this for each SDK.

Yes. Haven't had a chance to port to each language, but would love to have more examples to show.